ios/Classes/PushNotificationEventHandlers.swift

```diff
@@ -59,9 +59,33 @@ public class PushNotificationEventHandlers: NSObject, UNUserNotificationCenterDe
         if (application.applicationState == .background || application.applicationState == .inactive) {
             methodName = AblyPlatformMethod_pushOnBackgroundMessage
         }
-        let remoteMessage = RemoteMessage(data: userInfo._bridgeToObjectiveC(), notification: nil);
+        
+        let notification: Notification? = createNotification(from: userInfo);
+        let remoteMessage = RemoteMessage(data: userInfo._bridgeToObjectiveC(), notification: notification);
         methodChannel.invokeMethod(methodName, arguments: remoteMessage) { flutterResult in
             completionHandler(.newData);
         }
     }
+    
+    private func createNotification(from userInfo: [AnyHashable : Any]) -> Notification? {
+        let notification = Notification();
+        
+        if let aps = userInfo["aps"] as? NSDictionary {
+            if let alert = aps["alert"] as? NSDictionary {
+                if let title = alert["title"] as? String {
+                    notification.title = title
+                }
+                if let body = alert["body"] as? String {
+                    notification.body = body
+                }
+            }
+        }
+        
+        if (notification.title != nil || notification.body == nil) {
+            // Return an instance of notification only if it has some data.
+            return notification;
+        }
+        
+        return nil;
+    }
```

Comment:
how about this?
```swift
class Notification: NSObject {
  @objc
  var title: String?
  
  @objc
  var body: String?
  
  init?(title: String?, body: String?) {
    guard title != nil || body != nil else {
      return nil
    }
    
    self.title = title
    self.body = body
  }
}

// And the method

private func createNotification(from userInfo: [AnyHashable : Any]) -> Notification? {
  guard  let alert = userInfo["aps"] as? [AnyHashable: Any] else {
    return nil
  }
  
  return Notification(
    title: alert["title"] as? String, 
    body: alert["body"] as? String
  )
}

```

Reply:
I wanted to only create a notification if either title and body is set. If title and body is both null, your code would still create an empty `notification`. TBH, this is a case which might never happen though, `aps` might always be set. What do you think?

Reply:
In my example if title is nil and body is nil the Notification object won’t be created, that’s what guard returns.

Reply:
I've tried to implement this in a slightly different way, and there I found an issue with using convenience initializers with **failable (returns nil)** initializer. The compiler wants me to call `self.init`:

```
class Notification: NSObject {
    @objc
    var title: String?;
    @objc
    var body: String?;
    
    init?(title: String?, body: String?) {
      guard title != nil || body != nil else {
        return nil
      }
      
      self.title = title
      self.body = body
    }
    
    convenience init?(from userInfo: [AnyHashable : Any]) {
        if let aps = userInfo["aps"] as? NSDictionary {
            if let alert = aps["alert"] as? NSDictionary {
                let title = alert["title"] as? String
                let body = alert["body"] as? String
                self.init(title: title, body: body)
            } else {
                self.init(title: nil, body: nil)
                return nil
            }
        } else {
            self.init(title: nil, body: nil)
            return nil
        }
    }
}
```

- Notice the ugly usage of `self.init(title: nil, body: nil)` and `return nil`. Without `return nil`, we would return the instance.
- I cannot return `self` to finish this initializer early, which would have made it cleaner. 
![CleanShot 2021-11-18 at 08 46 30](https://user-images.githubusercontent.com/24711048/142381872-e392c38a-527e-46ee-bfa4-035061a2e72c.png)

Reply:
Well, you can't use return `self` in swift initialisers, if it's optional what you can return is just `nil`. So remove `return self` and change `self.init(title: nil, body: nil)` to `return nil` and it should be OK

Reply:
and please change casting from `NSDictionary` (it's obj-c) to `[AnyHashable: Any]`

Reply:
I double checked if `title` is always available in userInfo.aps.alert, and it is not: See example 4 in https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/CreatingtheNotificationPayload.html#//apple_ref/doc/uid/TP40008194-CH10-SW1:
```
{
    "aps" : {
        "alert" : {
            "loc-key" : "GAME_PLAY_REQUEST_FORMAT",
            "loc-args" : [ "Jenna", "Frank"]
        },
        "sound" : "chime.aiff"
    },
    "acme" : "foo"
}
```

This happens when the user wants to use localization using iOS files (Localizable.strings file). This is not supported by Ably Flutter. See [Alert Keys](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/PayloadKeyReference.html#//apple_ref/doc/uid/TP40008194-CH17-SW1) for more information. 

So, in summary, it is safe to say if `alert` is set, the user should also set at least a `title`. 61bbbe22601998faf08b070a88ec92ba78376d41